execution/src/execution_impl.rs

```diff
@@ -12,80 +14,123 @@ use wasmer_middlewares::Metering;
 
 const MAIN: &str = "main";
 
-fn print(env: &Env, arg: i32) {
-    let str_ptr = StringPtr::new(arg as u32);
-    // May be in wasmer-as create a pointer that keep memory?
-    println!(
-        "\n{}",
-        str_ptr
-            .read(env.wasm_env.memory.get_ref().expect("initialized memory"))
-            .unwrap()
-    );
+/// Utility function to call a WASM module (compiled from AssemblyScript) using
+/// high-level Rust types (by de/serializing a String).
+///
+/// The intend is to be able to expose any kind of function as an ABI:
+///
+/// ```ignore
+/// fn get_status() -> Result<NodeStatus> {
+///     Ok(typed_call(env, "0xAddressOfSCInLedger", "get_status", ())?)
+/// }
+/// ```
+pub fn _typed_call<T: Serialize, R: DeserializeOwned>(
+    env: &Env,
+    address: &Address,
+    function: &str,
+    _args: T, // TODO: @adrien-zinger how do we pass function args in AssemblyScript?
+) -> Result<R> {
+    let value = call_module(env, address, function);
+    Ok(serde_json::from_str::<R>(&value.ret)?)
 }
 
-fn call(env: &Env, address: i32, function: i32) -> i32 {
-    let memory = env.wasm_env.memory.get_ref().expect("initialized memory");
-    let addr_ptr = StringPtr::new(address as u32);
-    let func_ptr = StringPtr::new(function as u32);
-    let address = &addr_ptr.read(memory).unwrap();
-    let fnc = &func_ptr.read(memory).unwrap();
-    type GmSign = fn(&Address) -> Result<Bytecode>;
-    let get_module: GmSign = env.interface.get_module;
+/// Call an exported function in a WASM module at a given address
+fn call_module(env: &Env, address: &Address, function: &str) -> Response {
+    let get_module: fn(&Address) -> Result<Bytecode> = env.interface.get_module;
     let module = &get_module(address).unwrap();
-    let value = exec(
+    exec(
         env.remaining_points,
         None,
         module,
-        fnc,
+        function,
         &vec![],
         &env.interface,
     )
-    .unwrap();
+    .unwrap() // TODO: Should return a Result<Response>
+}
+
+/// Raw call that have the right type signature to be able to be call a module
+/// directly form AssemblyScript:
+///
+/// - get_string.wat
+/// ```ts
+/// export function getString(): string {
+///   return "hello test";
+/// }
+/// ```
+///
+/// - caller.wat
+/// ```ts
+/// export declare function call_module(address: string, func:string): string
+/// export declare function print(message: string): void
+///
+/// export function main(): i32 {
+///   let string_from = call_module("get_string.wat", "getString")
+///   print(string_from)
+///   return 0;
+/// }
+/// ```
+fn assembly_script_call_module(env: &Env, address: i32, function: i32) -> i32 {
+    let memory = env.wasm_env.memory.get_ref().expect("initialized memory");
+    // TODO: replace all unwrap() by expect()
```

Comment:
why not return a result ?

Reply:
We couldn't change the signature of `assembly_script_` function (the ones that we used as `NativeFunction`)!